### PR TITLE
Add tutorial for HA with keepalived, Failover IPv4 and Cloud vLAN

### PR DESCRIPTION
**Pull request by pabstma**
This adds a tutorial for a (simple) high availability solution using keepalived, Netcup Failover IPv4, Netcup Cloud vLAN and two Netcup vServers or Root Servers. It is intended as a more advanced tutorial where not every step is ellaborated to the smallest detail.

---

I have read and understood the Contributor's Certificate of Origin at the end of the template and I hereby certify that I meet the contribution criteria described in it.

Signed-off-by: Marcel Pabst pabstmarcel95@gmail.com

### CODE REVIEW

**Comment by raphi-0901** [comment]
Hey, I´m seeing the packets sending with this command `sudo tcpdump proto 112 -i eth1`. Then I turn off my s1. Should I see in SCP any changes then?

File: community-tutorials/high-availability-keepalived/01-en.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+# Step 1 - Setup Cloud vLAN
+## Step 1.1 - Preparation SCP
+### Step 1.1.1 - Log in to your [SCP](https://www.servercontrolpanel.de), either directly or via the autologin feature in your [CCP](https://www.customercontrolpanel.de)
+### Step 1.1.2 - Choose a server (from the list on the top left) 
+### Step 1.1.3 - Navigate to the *Network* section (in the left sidebar)
+### Step 1.1.4 - Choose *add Ethernet Interface* (on the top right)
+### Step 1.1.5 - Select your Cloud vLAN and choose *virtio* as the driver
+### Step 1.1.6 - Select the second server and repeat the steps 1.1.3 - 1.1.5; then continue with 1.2
+## Step 1.2 - Create interface
+Log in to your servers and create a vLAN-configuration for the network interface via your favourite text editor e.g. `sudo vim /etc/network/interfaces.d/vlan.cfg`
```
Comment (raphi-0901) [inline community-tutorials/high-availability-keepalived/01-en.md:41]: I needed to switch back from netplan to network package in order to follow your tutorial. Netplan is preinstalled in Ubuntu. Did you have to do the same?

File: community-tutorials/high-availability-keepalived/01-en.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+One interesting parameter to start tweaking is the *advert\_int* parameter. This parameter on the one hand determines how often an advert should be sent (every 2 seconds in this case) and on the other hand determines that if after 3 * *advert\_int* seconds no advert was received the failover is triggered (so in this case; after 3 * 2 = 6 seconds).
+
+If a server changes from BACKUP to MASTER state the *notify\_master* script, which is our *to\_master.sh* (that reroutes our Failover IPv4), is triggered.
+
+But there are much more methods you can use to trigger a failover besides the adverts. You could for example also track files, scripts, processes, interfaces etc. to increase/decrease the priority and trigger the failover based on this. Especially scripts arev very flexible and offer the possibility for more sophisticated checks. If you are interested in more configuration methods this [Red Hat Article](https://www.redhat.com/sysadmin/advanced-keepalived) covers some configurations.
+
+**Example */etc/keepalived/keepalived.conf* for s2**
```
Comment (raphi-0901) [inline community-tutorials/high-availability-keepalived/01-en.md:173]: Should be `s1` right?

**Review by raphi-0901 — CHANGES_REQUESTED** [review]

File: community-tutorials/high-availability-keepalived/01-en.md (hunk 2 above)
Comment (pabstma) [reply community-tutorials/high-availability-keepalived/01-en.md:173]:
```suggestion
**Example */etc/keepalived/keepalived.conf* for s1**
```
Yes, that should be s1

File: community-tutorials/high-availability-keepalived/01-en.md (hunk 1 above)
Comment (pabstma) [reply community-tutorials/high-availability-keepalived/01-en.md:41]: Yes that's true. I will check that and add a corresponding netplan config or remove Ubuntu from the supported distros.

**Comment by pabstma** [comment]
> Hey, I´m seeing the packets sending with this command `sudo tcpdump proto 112 -i eth1`. Then I turn off my s1. Should I see in SCP any changes then?

If you followed the tutorial completely you should see the Failover IP assigned to s2 if s1 is not sending adverts anymore/or the adverts do not reach s2; instead of s1 which can be seen in the network section of s2 in the scp.

File: community-tutorials/high-availability-keepalived/01-en.md (hunk 1 above)
Comment (pabstma) [reply community-tutorials/high-availability-keepalived/01-en.md:43]: I removed Ubuntu from the supported distros and added a hint that the tutorial can be used as a starting point for other distributions besides Debian but may need modifications for that (https://github.com/netcup-community/community-tutorials/pull/49/commits/231e0971dfc837063dfb722a91bc1bf41bccc14a).

File: community-tutorials/high-availability-keepalived/01-en.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    interface eth1
+    virtual_router_id 101
+    priority 255
+    advert_int 2
+    authentication {
+        auth_type PASS
+        auth_pass CUSTOM_PASSWORD
+    }
+    virtual_ipaddress {
+        10.132.0.100
```
Comment (raphi-0901) [inline community-tutorials/high-availability-keepalived/01-en.md:188]: Which IP address are we referring here?

**Review by raphi-0901 — CHANGES_REQUESTED** [review]

**Comment by raphi-0901** [comment]
When I run `sudo journalctl -u keepalived` I get the following:
```
Nov 16 08:26:07 v220211184008137980 Keepalived_vrrp[759]: (VI_1) Entering BACKUP STATE (init)
Nov 16 08:26:13 v220211184008137980 Keepalived_vrrp[759]: (VI_1) Entering MASTER STATE
```
This happened right after restarting s2 while s1 was running. And it is not visible in SCP.

File: community-tutorials/high-availability-keepalived/01-en.md (hunk 3 above)
Comment (pabstma) [reply community-tutorials/high-availability-keepalived/01-en.md:188]: That is the virtual ip address of our vrrp master. So at any time either s1 or s2 owns this address. We currently do not use it but if we had other servers in our vlan we could also use this IP to connect to the master server.

**Comment by pabstma** [comment]
> When I run `sudo journalctl -u keepalived` I get the following:
> 
> ```
> Nov 16 08:26:07 v220211184008137980 Keepalived_vrrp[759]: (VI_1) Entering BACKUP STATE (init)
> Nov 16 08:26:13 v220211184008137980 Keepalived_vrrp[759]: (VI_1) Entering MASTER STATE
> ```
> 
> This happened right after restarting s2 while s1 was running. And it is not visible in SCP.

Please always refer to the server where the logs are from. 

After starting the service on both servers you should see s1 entering the master state `Nov 16 09:04:08 v2202104129099154891 Keepalived_vrrp[1843355]: (VI_1) Entering MASTER STATE` and s2 entering backup state `Nov 16 09:03:58 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Entering BACKUP STATE (init)`

If you restart s1 now (which is our master) then you should see something like this on s2:
```
Nov 16 09:09:00 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Backup received priority 0 advertisement
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Receive advertisement timeout
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Entering MASTER STATE
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) setting VIPs.
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Sending/queueing gratuitous ARPs on eth1 for 10.132.0.100
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:01 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:06 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Sending/queueing gratuitous ARPs on eth1 for 10.132.0.100
Nov 16 09:09:06 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:06 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:06 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:06 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
Nov 16 09:09:06 v2202104129099154895 Keepalived_vrrp[464409]: Sending gratuitous ARP on eth1 for 10.132.0.100
```
We receive a priority 0 advertisement and an advertisement timeout, switching to master state right after that, assigning the VIP to ourself and sending some ARP packages. What happens in the background is the `notify_master`script is called and reroutes our Failover IP via a call to the Netcup Webservice. Only the change of the Failover IP can be seen in the SCP nothing else.

When s1 comes online again s2 switches back to backup which should look something like this in the logs of s2:
```
Nov 16 09:11:03 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Master received advert from 10.132.0.10 with higher priority 255, ours 254
Nov 16 09:11:03 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) Entering BACKUP STATE
Nov 16 09:11:03 v2202104129099154895 Keepalived_vrrp[464409]: (VI_1) removing VIPs.
```
And the whole switching of the Failover IP to s2 and back to s1 looks like this in the SCP logs of s2:
Screenshot_20211116_091657

**Comment by raphi-0901** [comment]
The logs came from s2.
S2 enters the master state 6 seconds after rebooting although s1 is running. I guess I mocked something up with the script.

**Comment by pabstma** [comment]
> The logs came from s2. S2 enters the master state 6 seconds after rebooting although s1 is running. I guess I mocked something up with the script.

Alright, s2 shouldn't be master when s1 is running and it receives the adverts from s1 (which you already checked with `sudo tcpdump proto 112 -i eth1` on s2). Maybe some configuration error in the keepalived config? Make sure s2 has a lower priority than s1 and they have the same virtual_router_id. I checked the example keepalived configurations in the tutorial again and they look correct to me.

**Comment by raphi-0901** [comment]
I tried it again with Debian (I was still using Ubuntu) and now it works. 👍

**Comment by raphi-0901** [comment]
> I tried it again with Debian (I was still using Ubuntu) and now it works. 👍

I forgot to check if the changes are visible in SCP and they are not. But when I run `sudo journalctl -u keepalived` I get the same result as you.

**Comment by pabstma** [comment]
> 
> 
> > I tried it again with Debian (I was still using Ubuntu) and now it works. 👍
> 
> I forgot to check if the changes are visible in SCP and they are not. But when I run `sudo journalctl -u keepalived` I get the same result as you.

Alright, so keepalived itself works as intended but the routing of the Failover IP is not. The routing of the Failover IP is changed by a call to the Netcup Webservice via our `to_master.sh` script. Could you please check if `curl` is installed on your system (if not `sudo apt install curl`) and then try again. Please also make sure that the `keepalived.conf` contains the correct path to the `to_master.sh`script in the `notify_master /PATH/TO/YOUR/SCRIPT/to_master.sh` section and that the script has execution rights.

**Comment by raphi-0901** [comment]
I found the error. My password for the webservice contained a `<` which broke the whole script.

**Comment by netcupClaudiaM** [comment]
Hi pabstma, thank you for contributing! It's been a while since you opened this PR and it probably is an open secret we're currently overwhelmed with reviewing all the awesome tutorials we got. Please  be patient - we're about to publish tutorials this and next week. Thank you!

**Pull request merged**